UI - fix encoding for user-entered paths

meirish opened the pull request:
By default ember will encode and decode dynamic route segments for you, but there are several places where we're using star dynamic segments which _don't_ come with this behavior so we need to do that encoding and decoding ourselves.

This is also true in ember data - model ids are encoded when you pass them to ember data's `buildUrl` method. There are several places where we're constructing api URLs via string concatenation, and not properly encoding the user-entered items. 

Additionally, when passing the path into the filter box on the secret list page, we were not escaping the path so that we could create a valid Regex object. This meant that any characters that were valid in Regex pattern matching being present in the paths could potentially cause an error.

This PR changes the app to make sure we're encoding when passing elements to star segment urls, decoding when we're pulling params from those same routes, and properly escaping strings before passing them to the `Regex` constructor in the list-controller mixin.

Fixes #6282 .

johncowen commented on ui/app/controllers/vault/cluster/secrets/backend/list.js:
```diff
@@ -4,50 +4,19 @@ import Controller from '@ember/controller';
-    setFilterFocus(bool) {
-      this.set('filterFocused', bool);
-    },
-
```
Lots of removals here, I guess this is all covered in `ListController` now?

meirish replied:
Yep! It was previously, we just weren't using the mixin very much.

johncowen commented on ui/app/adapters/secret-v2-version.js:
```diff
@@ -3,13 +3,14 @@ import { isEmpty } from '@ember/utils';
   namespace: 'v1',
   _url(backend, id, infix = 'data') {
-    let url = `${this.buildURL()}/${backend}/${infix}/`;
+    let url = `${this.buildURL()}/${encodePath(backend)}/${infix}/`;
```
Ah this was the other possibly user generated value here `infix`. I'm 99.9% sure it's probably hardcoded somewhere else I can see here. But just to triple sanity check with you in case its coming from a user generated value somewhere.

meirish replied:
Yep, it's hardcoded - it'll either be `data` or `metadata` - the apis for v2 have that prefix (infix here because there's /v1/backend_path at the start of all of them).

johncowen commented on ui/app/adapters/secret-engine.js:
```diff
@@ -38,14 +39,14 @@ export default ApplicationAdapter.extend({
     if (query.type === 'aws') {
-      return this.ajax(`/v1/${query.backend}/config/lease`, 'GET').then(resp => {
+      return this.ajax(`/v1/${encodePath(query.backend)}/config/lease`, 'GET').then(resp => {
         resp.path = query.backend + '/';
```
Again just a sanity check that this doesn't need encoding here, you encode it on 49 but not on 50

meirish replied:
Yeah we're only encoding when it goes to the server via ajax, and when it goes to the URL for routing - here it's going to be put in ember data for the model so we want the decoded version for display.

johncowen commented on ui/app/adapters/role-jwt.js:
```diff
@@ -9,7 +10,7 @@ export default ApplicationAdapter.extend({
-    let url = `/v1/auth/${path}/oidc/auth_url`;
+    let url = `/v1/auth/${encodePath(path)}/oidc/auth_url`;
     let redirect_uri = `${window.location.origin}${this.router.urlFor('vault.cluster.oidc-callback', {
       auth_path: path,
```
Little sanity check, line 13 you encode line 15 you don't. 99.9% sure thats ok, but just to check.

meirish replied:
Oh yep, urlFor is going to want encoded since it's router-related - thanks!

meirish commented on ui/app/adapters/transit-key.js:
```diff
@@ -47,29 +48,29 @@ export default ApplicationAdapter.extend({
     if (action === 'export' && param) {
       let [type, version] = param;
-      const exportBase = `${urlBase}/${type}-key/${id}`;
+      const exportBase = `${urlBase}/${type}-key/${encodePath(id)}`;
```
Yep, the type and action are hard-coded elsewhere in the app, the only thing the user does is name the transit key (which is the id).

Good spot on the array vs string, I'll follow up on that!

Re: `const`/`let` - yeah we need to decide as a team how we want to use this. I personally don't think const adds much most of the time (which is also why it's all over the place across the codebase), but we should decide on a standard and stick to it.

meirish replied:
Ok, found the `param` bit - looks like in export it is an array, and datakey it's a string - I created a ticket internally to track that as we have upcoming work in that area which would be a good time to refactor it.

meirish commented on ui/app/utils/args-tokenizer.js:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  return args;
+}
```
I have some changes I need to push - we went back to yargs-parser - but the change here had to do with handling " and ' in the web console - the tokenizer tries to be smart about where splitting should occur based on quotes and single / double quote break it. I'm thinking maybe we shouldn't support single or double quotes in paths in the web cli, but haven't landed anywhere yet.

The vars are from the source file - https://github.com/yargs/yargs-parser/blob/master/lib/tokenize-arg-string.js

meirish commented on ui/app/utils/path-encoding-helpers.js:
```diff
@@ -0,0 +1,14 @@
+
+const {
+  Normalizer: { normalizePath, encodePathSegment },
+} = RouteRecognizer;
```
https://github.com/tildeio/route-recognizer/blob/master/lib/route-recognizer/normalizer.ts#L20 has some good comments - I was thinking it'd be nice to have the same semantics as the router. I think for adapter use it's not strictly necessary, but decoded slashes still look nicer.

meirish commented on ui/package.json:
```diff
@@ -106,13 +107,14 @@
     "walk-sync": "^0.3.3",
     "xstate": "^3.3.3",
-    "yargs-parser": "^10.0.0"
+    "yargs-parser": "^12.0.0"
```
We tried upgrading first, then pulled in the file we were using - but this has changed in the latest back to 10.

meirish commented on ui/tests/acceptance/secrets/backend/kv/secret-test.js:
```diff
@@ -259,4 +259,44 @@ module('Acceptance | secrets/secret/create', function(hooks) {
+      ']',
+      '^',
+      '_',
+    ].map(char => `${char}some`);
```
😄 - we wanted to recreate the script from the reported issue, and doing it this way meant we got to test the web cli too (and found it buggy) so win, win!

meirish commented:
@johncowen Thanks for checking this out! I think I addressed all of the points you raised. The tests here just cover secret stuff as that's where Vault server is most lenient in allowed paths. I dug into the error we saw when I tried to create a transit key named `#foo` and it turns out that many endpoints in vault are more strict wrt characters, so that's disallowed at the server level.

johncowen commented:
Ah yeah course! Makes sense. Good stuff, anyway thanks for the walkthrough and chat also, good to catch up!

John

dupainaulevain commented:
@meirish thanks for the fix!
